docs:  add peas and pod (#182)
jina-ai/docs: chapters/peas-and-pods.rst

Development Guide: Peas and Pods in Jina
=========================================

.. meta::
   :description: Development Guide: Peas and Pods in Jina
   :keywords: Jina, pea, pod

.. note:: This guide assumes you have a basic understanding of Jina, if you haven't, please check out `Jina 101 <https://101.jina.ai>`_ first.
.. note:: This guide assumes you have a basic understanding of Parallelization inside Jina, if you haven't, please check out `Parallelization in Jina <../parallel>`_ first.

You might have already learned from `Jina 101 <../101.rst>`_.

> Jina :term:`Pea` wraps a :term:`Driver`, :term:`Executor` and lets it exchange data with other Peas.

Reviewer: wraps a driver and an Exectuor? Why ,

Author: removed driver

Jina :term:`Pod` is a context manager for one or multiple Peas that have the same properties.
It coordinates Peas to improve efficiency and scaling.
Beyond that, a Pod adds further control, scheduling, and context management to its Peas.

In this tutorial, we dive deeper into Peas and Pods,
to help you gain in-depth knowledge on the architecture and design principle of Peas and Pods in Jina.

.. contents:: Table of Contents
    :depth: 3

Pea and Pod Explained
^^^^^^^^^^^^^^^^^^^^^^^

Singleton Or Not?
------------------

> In jina, if a Pod only contains one Pea, we call it a Singleton Pod.

Reviewer: Jina

> In jina, if a Pod only contains one Pea, we call it a Singleton Pod.

Reviewer: only one Pea

While in the second case, when user specifically defined a `parallel` value greater than 1,
Jina will add a :term:`HeadPea` and :term:`TailPea` to the same Pod.
HeadPea distribute traffic to different Peas inside the Pod,
TailPea collects the calculated result from the Peas inside the Pod.

Let's say we defined `parallel=3` inside a Pod,
then we have 5 Peas inside the Pod.
Traffic first arrives HeadPea, and the HeadPea distributes traffic to the 3 Peas in parallel,
and finally the result will be sent to the TailPea.

Jina offers parallelization at different levels, such as

1. Local Threads/Process with `BasePea`.
2. Inside container with `BasePea` inside `ContainerRuntime`.
3. Remotely with `RemotePeas` and `RemotePods` inside `JinadRuntime`.

Jina Daemon (JinaD) enables Jina to spin up and distribute Peas, Pods, Flows in any system.

Stateless vs Stateful Pea
---------------------------

In the previous section we mentioned `parallel`.
This parameter manages different Peas inside a Pod,
in a Stateless fashion.
Since a Pea is a wrap over Jina Executor.
Dependent on the type of the Executor, we consider a Pea is Stateless or Statefull.

> If a Pea wraps Encoder, Crafter, Segmenter, Ranker, Evaluator, Classifier, we consider it as Stateless.

Reviewer: wraps an

> Since this Pea do not persist data.

Reviewer: will not persist data

> On the other hand, if a Pea wraps Indexer, we consider it as Stateful.

Reviewer: wraps an Indexer

Because we persist indexed result to our workspace.

We use `parallel` to manage the number of Stateless Peas inside a Pod.
We use `shards` to manage the number of Stateful Peas inside a Pod.

How Local Pea and Remote Pea interact with Pod
-----------------------------------------------

Let's first think about the case of a Singleton Pod runs locally.
This Pod will only have 1 Pea inside, i.e. `parallel=1` or `shard=1`,
depends on the type of Executor it wraps.
The Pea inside the previous Pod will receive traffic from the last Pea,
process the data and send data to the next Pea.

> When it comes to a Singleton Pod runs remotely (i.e. runs with JinaD),

Reviewer: running remotely

the main process will setup `JinadRuntime`,
the runtime is going to create remote Peas inside the remote machine.
The JinadRuntime take ownership of 2 tasks:

1. Parse args to JinaD to fire up a remote Pod with a single Pea using `put/pod` api endpoint.
2. Streaming logs from the remote Pod to the local Pod.

The remote Pea inside the remote Pod will process the data,
the result will be send to the next Pea.
In the next section we will talk about the most complex scenario: "distributed Remote Peas in Pod".

Distributed Remote Peas in Pod
-------------------------------

In this section, we talk about how multiple Peas (`parallel` or `shards` greater or equal to 2) runs inside a remote Pod.
First of all, the main process will setup `JinadRuntime`.
`JinadRuntime` create a new remote Pod inside the remote machine, with a HeaderPea,
a TailPea and `1` Pea inside the same machine,
the host of the Pea will stay the same as the Pod.

Take a look at the image below.
Imaging we defined `parallel=2` in our Flow/YAML configuration.

> Since the current remote machine only has 1 ip address (`host=0.0.0.3`),

Reviewer: has one IP address 
(numbers one to ten should normally be written out, anything more then ten can be a number). 

IP should be uppercase

the Jina main process will setup another `JinadRuntime`,

> and fire up another Pea inside the 2nd host (`0.0.0.4`),

Reviewer: second host

> by calling `put/pea` api endpoint.

Reviewer: API endpoint

Development Guide: Peas and Pods in Jina
=========================================

.. meta::
   :description: Development Guide: Peas and Pods in Jina
   :keywords: Jina, pea, pod

.. note:: This guide assumes you have a basic understanding of Jina, if you haven't, please check out `Jina 101 <https://101.jina.ai>`_ first.
.. note:: This guide assumes you have a basic understanding of Parallelization inside Jina, if you haven't, please check out `Parallelization in Jina <../parallel>`_ first.

> You might already learned from `Jina 101 <../101.rst>`_.

Reviewer 2: ```suggestion
You might have already learned from `Jina 101 <../101.rst>`_.
```

> Jina :term:`Pea` wraps an :term:`Executor` and lets it exchange data with other Peas.

Reviewer 3: `Pea` wraps a `Driver`, `Executor`.

> Jina :term:`Pod` is a container and interface for one or multiple Peas that have the same properties.

Reviewer 3: is a `context manager` maybe better?

It coordinates Peas to improve efficiency and scaling.
Beyond that, a Pod adds further control, scheduling, and context management to its Peas.

In this tutorial, we dive deeper into Peas and Pods,

> to help you gain in depth knowledge on the architecture and design principle of Peas and Pods in Jina.

Reviewer 2: ```suggestion
to help you gain in-depth knowledge on the architecture and design principle of Peas and Pods in Jina.
```

.. contents:: Table of Contents
    :depth: 3

Pea and Pod Explained
^^^^^^^^^^^^^^^^^^^^^^^

Singleton Or Not?
------------------

In jina, if a Pod only contains one Pea, we call it a Singleton Pod.

> If a user, define `parallel` as 1, then it will be considered as Singleton.

Reviewer 2: ```suggestion
If a user defines `parallel` as 1, then it will be considered Singleton.
```

> Otherwise, one Pod manages at least two Peas. e.g. `parallel` is at least 2.

Reviewer 2: Put reference to the existing docs on parallelism in Jina

Author: @cristianmtr referenced at the top of the blog: 
![image](https://user-images.githubusercontent.com/9794489/111574720-4e703300-87ad-11eb-82a5-9033d88fc4ab.png)

In the first case, a Pod only manages a single Pea.

> While in the second case, when user specifically defined a `parallel` value greater equal than 2,

Reviewer 2: ```suggestion
While in the second case, when user specifically defined a `parallel` value greater than 1,
```

Jina will add a :term:`HeadPea` and :term:`TailPea` to the same Pod.
HeadPea distribute traffic to different Peas inside the Pod,
TailPea collects the calculated result from the Peas inside the Pod.

Let's say we defined `parallel=3` inside a Pod,
then we have 5 Peas inside the Pod.

> Traffic firstly arrive HeadPea, and the HeadPea distribute traffic two 3 peas in Parallel,

Reviewer 2: ```suggestion
Traffic first arrives HeadPea, and the HeadPea distributes traffic to the 3 Peas in parallel,
```

> finally the result will be send to TailPea.

Reviewer 2: ```suggestion
and finally the result will be sent to the TailPea.
```

> Jina offers parallelization at different levels, such as

Reviewer 2: This can be confused with intra- and inter-parallelism, as discussed in the parallelism chapter. Make sure clarify

Author: @cristianmtr referenced at the top of the blog, guess I don't need to reference for a 2nd time?
![image](https://user-images.githubusercontent.com/9794489/111574486-ddc91680-87ac-11eb-8923-bc568b7158de.png)

Author: added a note below

> 1. Local Threads/Process with `BasePea`.

Reviewer 2: This is covered in the exists docs on parallelism. Don't duplicate info, as it will be harder to maintain

2. Inside container with `BasePea` inside `ContainerRuntime`.
3. Remotely with `RemotePeas` and `RemotePods` inside `JinadRuntime`.

> Jina Daemon (JinaD) enable Jina to spin up and distribute Peas, Pods, Flows in any system.

Reviewer 3: ```suggestion
Jina Daemon (JinaD) enables Jina to spin up and distribute Peas, Pods, Flows in any system.
```

Stateless vs Stateful Pea
---------------------------

> In the previous section we mentioned `parallel`.

Reviewer 2: Again, covered in the other chapter.

Author: @cristianmtr I'd like to keep it here for 2 reasons:

1. This is a documentation on Pea and Pod, I think it's reasonable to leave it here.
2. This part will be needed (better understanding) by the later section: distributed peas in pod.

I personally think a bit of overlap is not a big issue.

This parameter manages different Peas inside a Pod,
in a Stateless fashion.
Since a Pea is a wrap over Jina Executor.
Dependent on the type of the Executor, we consider a Pea is Stateless or Statefull.

If a Pea wraps Encoder, Crafter, Segmenter, Ranker, Evaluator, Classifier, we consider it as Stateless.
Since this Pea do not persist data.
On the other hand, if a Pea wraps Indexer, we consider it as Stateful.
Because we persist indexed result to our workspace.

We use `parallel` to manage the number of Stateless Peas inside a Pod.
We use `shards` to manage the number of Stateful Peas inside a Pod.

How Local Pea and Remote Pea interact with Pod
-----------------------------------------------

Let's first think about the case of a Singleton Pod runs locally.
This Pod will only have 1 Pea inside, i.e. `parallel=1` or `shard=1`,
depends on the type of Executor it wraps.

> The Pea inside the Pod will receive traffic from the last Pea,

Reviewer 3: `previous` Pod?

process the data and send data to the next Pea.

When it comes to a Singleton Pod runs remotely (i.e. runs with JinaD),
the main process will setup `JinadRuntime`,

> the runtime is going to create a new Pod (Remote) inside the remote machine.

Reviewer 3: will create remote `Peas` and not `hosts`

2. Streaming logs from the remote Pod to the local Pod.

The remote Pea inside the remote Pod will process the data,
the result will be send to the next Pea.
In the next section we will talk about the most complex scenario: "distributed Remote Peas in Pod".

Distributed Remote Peas in Pod
-------------------------------

In this section, we talk about how multiple Peas (`parallel` or `shards` greater or equal to 2) runs inside a remote Pod.
First of all, the main process will setup `JinadRuntime`.
`JinadRuntime` create a new remote Pod inside the remote machine, with a HeaderPea,
a TailPea and `1` Pea inside the same machine,
the host of the Pea will stay the same as the Pod.

Take a look at the image below.
Imaging we defined `parallel=2` in our Flow/YAML configuration.
Since the current remote machine only has 1 ip address (`host=0.0.0.3`),
the Jina main process will setup another `JinadRuntime`,
and fire up another Pea inside the 2nd host (`0.0.0.4`),
by calling `put/pea` api endpoint.

.. image:: ./images/remote-peas-in-pod.png
  :width: 600
  :alt: remote peas in pod
  :align: center

Then the HeadPea is going to distribute traffic to both Peas.

> In the end, the TailPeae collects data from Pea1 and Pea2,

Reviewer 3: typo in TailPeae

Development Guide: Peas and Pods in Jina
=========================================

.. meta::
   :description: Development Guide: Peas and Pods in Jina
   :keywords: Jina, pea, pod

.. note:: This guide assumes you have a basic understanding of Jina, if you haven't, please check out `Jina 101 <https://101.jina.ai>`_ first.
.. note:: This guide assumes you have a basic understanding of Parallelization inside Jina, if you haven't, please check out `Parallelization in Jina <../parallel>`_ first.

You might have already learned from `Jina 101 <../101.rst>`_.
Jina :term:`Pea` wraps an :term:`Executor` and lets it exchange data with other Peas.

> Jina :term:`Pod` is a context manager for one or multiple Peas that have the same properties.

Reviewer 3: I would add something on these lines.

A Pod is an abstraction layer on top of Peas equipped with the same Executors allowing the user to reason about the application and design its functionality hiding the parallelization and the non-functional details

Author: added @JoanFM

to help you gain in-depth knowledge on the architecture and design principle of Peas and Pods in Jina.

.. contents:: Table of Contents
    :depth: 3

Pea and Pod Explained
^^^^^^^^^^^^^^^^^^^^^^^

Singleton Or Not?
------------------

In Jina, if a Pod only contains only one Pea, we call it a Singleton Pod.
If a user, define `parallel` as 1, then it will be considered as Singleton.
Otherwise, one Pod manages at least two Peas. e.g. `parallel` is at least 2.

In the first case, a Pod only manages a single Pea.
While in the second case, when user specifically defined a `parallel` value greater than 1,
Jina will add a :term:`HeadPea` and :term:`TailPea` to the same Pod.
HeadPea distribute traffic to different Peas inside the Pod,
TailPea collects the calculated result from the Peas inside the Pod.

Let's say we defined `parallel=3` inside a Pod,
then we have 5 Peas inside the Pod.
Traffic first arrives HeadPea, and the HeadPea distributes traffic to the 3 Peas in parallel,
and finally the result will be sent to the TailPea.

Jina offers parallelization at different levels, such as

.. note::
   The parallelization here refers to intra-parallelism.

>    Checkout `this page <../parallel/#inter-parallelism>`_ to discover the difference between inter-parallelism and intra-parallelism in Jina.

Reviewer 2: ```suggestion
   Check out `this page <../parallel/#inter-parallelism>`_ to discover the difference between inter-parallelism and intra-parallelism in Jina.
```

1. Local Threads/Process with `BasePea`.
2. Inside container with `BasePea` inside `ContainerRuntime`.
3. Remotely with `RemotePeas` and `RemotePods` inside `JinadRuntime`.

Jina Daemon (JinaD) enables Jina to spin up and distribute Peas, Pods, Flows in any system.

Stateless vs Stateful Pea
---------------------------

In the previous section we mentioned `parallel`.
This parameter manages different Peas inside a Pod,

> in a Stateless fashion.

Reviewer 2: Why are `Stateless` and `Statefull` in uppercase?

Author: updated

the result will be send to the next Pea.
In the next section we will talk about the most complex scenario: "distributed Remote Peas in Pod".

Distributed Remote Peas in Pod
-------------------------------

In this section, we talk about how multiple Peas (`parallel` or `shards` greater or equal to 2) runs inside a remote Pod.
First of all, the main process will setup `JinadRuntime`.
`JinadRuntime` create a new remote Pod inside the remote machine, with a HeaderPea,
a TailPea and `1` Pea inside the same machine,
the host of the Pea will stay the same as the Pod.

Take a look at the image below.
Imaging we defined `parallel=2` in our Flow/YAML configuration.
Since the current remote machine only has one IP address (`host=0.0.0.3`),
the Jina main process will setup another `JinadRuntime`,
and fire up another Pea inside the second host (`0.0.0.4`),
by calling `put/pea` API endpoint.

.. image:: ./images/remote-peas-in-pod.png
  :width: 600
  :alt: remote peas in pod
  :align: center

Then the HeadPea is going to distribute traffic to both Peas.
In the end, the TailPea collects data from Pea1 and Pea2,
and send data to the next Pea.

An Example
^^^^^^^^^^^^

> In the code snippet below, we demonstrate a flow with three Pods.

Reviewer 2: ```suggestion
In the code snippet below, we demonstrate a Flow with three Pods.
```

Singleton Or Not?
------------------

<<<<<<< HEAD
In Jina, if a Pod only contains only one Pea, we call it a Singleton Pod.
If a user, define `parallel` as 1, then it will be considered as Singleton.
=======
In jina, if a Pod only contains one Pea, we call it a Singleton Pod.
If a user defines `parallel` as 1, then it will be considered Singleton.

> >>>>>>> c9b3a9eae8ecdfc8436fe745c247f2af31ae383c

Reviewer 3: check this

Author: @JoanFM refresh.. it was outdated
